agent/xds/config.go

```diff
@@ -28,10 +28,15 @@ type ProxyConfig struct {
 	LocalConnectTimeoutMs int `mapstructure:"local_connect_timeout_ms"`
 
 	// Protocol describes the service's protocol. Valid values are "tcp",
-	// "http" and "grpc". Anything else is treated as tcp. The enables protocol
-	// aware features like per-request metrics and connection pooling, tracing,
-	// routing etc.
+	// "http" and "grpc". Anything else is treated as tcp. This enables
+	// protocol aware features like per-request metrics and connection
+	// pooling, tracing, routing etc.
 	Protocol string `mapstructure:"protocol"`
+
+	// BindAddress overrides the address the proxy's listener binds to. This
+	// enables proxies in network namespaces to bind to a different address
+	// than the host address.
+	BindAddress string `mapstructure:"bind_address"`
```

Comment:
Is this meant to be a `address:port`? Or only the address? Maybe we should clarify. We have inconsistent history in Consul of `addr:port` vs separate configs for them.

But for example the built-in proxy has separate `bind_address` and `bind_port` so it's probably best to match that?

Reply:
Hmm it looks like this is just an address which is OK should we add port too? I was actually under the impression that it was port that Nomad needed control over to make this dance work rather than just the IP?

Reply:
We do something similar for gateways where we allow configuring bind addresses. There is a new ServiceAddress type for containing a split apart host/ip and port so that could be used here instead of a string

Reply:
https://github.com/hashicorp/consul/blob/3eb3ee5a1577f85ca55f3e373766af51a37b501f/agent/xds/config.go#L56-L73

Reply:
I added BindPort in 38b1465 as the `Bind{Address,Port}` fields are closely related to the `Service.{Address,Port}` and `Proxy.LocalService{Address,Port}` fields, both of which split the IP and Port into 2 fields.

I did this before noticing your comment on `ServiceAddress` @mkeeler. Should I switch to that new struct? It's all the same to me; just want to do what's most idiomatic.

Reply:
So the reason I created that struct is when I added Tagged Addressing for services like we had for nodes. I didn't want to parse address:port all over the place so I just created a struct that can be decoded wherever we need it.

Personally I like the address and port being in the same struct so their interdependence is explicit. It also matches well with how Envoy actually structures the config.

Reply:
The argument for separate (other than consistency with other proxies) is that 99% of the time you only want to set the port and not the address and want the address to default to something sensible like 0.0.0.0 or the agent's IP. That's possible if you accept `:1234`. And vice versa - in some case the port is auto assigned or taken from the `port` field but you want to be able to control the address without specifying a port. It's marginally cleaner to do those things separately.

Consuls bind config is separate too.

Honestly I don't care that much since we have precedence both ways but if it's already built with BindPort sepearate I'd be up for leaving it that way for now.